### PR TITLE
ChargeLinks update and stop requests doesn't return reject bug is fixed

### PR DESCRIPTION
**Pull request by HenrikSommer**
## Description

ChargeLinks with an open-ended end date is persisted with a fixed maximum date. The temporary _ChargeLinks update and stop is not yet supported_-rule did not take this into account, which is now fixed.

A few minor unrelated improvements without functional impact are included.

## References

* #1087

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRule.cs [hunk 1]
```diff
@@ -41,16 +42,18 @@ public class ChargeLinksUpdateNotYetSupportedRule : IValidationRuleWithExtendedD
...
-                }
-            }
-
-            return true;
+            return _existingChargeLinks.All(existingLink => !IsOverlapping(existingLink, newLink));
+        }
+
+        private static bool IsOverlapping(ChargeLink existingLink, ChargeLinkDto newLink)
+        {
+            var newLinkEndDateTime = newLink.EndDateTime.TimeOrEndDefault();
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRule.cs:50]: I'd prefer handling the null rather that calculating this somewhat arbitrary helper value.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRuleTests.cs [hunk 2]
```diff
@@ -48,17 +47,19 @@ public void IsValid_WhenCalledWithValidChargeLinks_ReturnsTrue(string meteringPo
...
-        [InlineAutoMoqData(12, 20)]
-        [InlineAutoMoqData(20, null!)]
-        [InlineAutoMoqData(1, 30)]
-        public void IsValid_WhenCalledWithOverlappingChargeLinks_ReturnsFalse(int startDateDayOfMonth, int? endDateDayOfMonth, string meteringPointId, DocumentDto document)
+        [InlineAutoMoqData("2022-01-01T23:00:00Z", "2022-01-12T23:00:00Z", "2022-01-11T23:00:00Z", "2022-01-21T23:00:00Z")] // Intersects at beginning of existing
+        [InlineAutoMoqData("2022-01-12T23:00:00Z", "2022-01-20T23:00:00Z", "2022-01-11T23:00:00Z", "2022-01-21T23:00:00Z")] // Intersects at end of existing
+        [InlineAutoMoqData("2022-01-01T23:00:00Z", "2022-01-30T23:00:00Z", "2022-01-11T23:00:00Z", "2022-01-21T23:00:00Z")] // Encapsulates existing
+        [InlineAutoMoqData("2022-01-11T23:00:00Z", "2022-01-21T23:00:00Z", "2022-01-11T23:00:00Z", "2022-01-21T23:00:00Z")] // Exact match
+        [InlineAutoMoqData("2022-01-01T23:00:00Z", "9999-12-31T23:59:59Z", "2022-01-11T23:00:00Z", "9999-12-31T23:59:59Z")] // EndDate is DateTime.Max
+        [InlineAutoMoqData("2022-01-01T23:00:00Z", null!, "2022-01-11T23:00:00Z", "9999-12-31T23:59:59Z")] // EndDate is null!
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRuleTests.cs:56]: Is this the only relevant test case with a null enddate?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRuleTests.cs [hunk 3]
```diff
@@ -48,17 +47,19 @@ public void IsValid_WhenCalledWithValidChargeLinks_ReturnsTrue(string meteringPo
...
 
         [Theory]
-        [InlineAutoMoqData(1, 12)]
-        [InlineAutoMoqData(12, 20)]
-        [InlineAutoMoqData(20, null!)]
-        [InlineAutoMoqData(1, 30)]
-        public void IsValid_WhenCalledWithOverlappingChargeLinks_ReturnsFalse(int startDateDayOfMonth, int? endDateDayOfMonth, string meteringPointId, DocumentDto document)
+        [InlineAutoMoqData("2022-01-01T23:00:00Z", "2022-01-12T23:00:00Z", "2022-01-11T23:00:00Z", "2022-01-21T23:00:00Z")] // Intersects at beginning of existing
+        [InlineAutoMoqData("2022-01-12T23:00:00Z", "2022-01-20T23:00:00Z", "2022-01-11T23:00:00Z", "2022-01-21T23:00:00Z")] // Intersects at end of existing
+        [InlineAutoMoqData("2022-01-01T23:00:00Z", "2022-01-30T23:00:00Z", "2022-01-11T23:00:00Z", "2022-01-21T23:00:00Z")] // Encapsulates existing
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRuleTests.cs:53]: And how about the opposite (true subset)?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRule.cs (hunk 1 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRule.cs:50]: It's not really arbitrary, it is the same extension method we use to set the end date on the `ChargeLink` domain object in `ChargeLinkFactory`.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRuleTests.cs (hunk 3 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRuleTests.cs:53]: Yes, I will extend the `IsValid_WhenCalledWithValidChargeLinks_ReturnsTrue` test to test more valid cases.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRuleTests.cs (hunk 2 above)
Comment (HenrikSommer) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRuleTests.cs:56]: You're right, there are more, depending on the start date.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRule.cs [hunk 4]
```diff
@@ -41,16 +42,16 @@ public class ChargeLinksUpdateNotYetSupportedRule : IValidationRuleWithExtendedD
...
-            }
-
-            return true;
+            return _existingChargeLinks.All(existingLink => !IsOverlapping(existingLink, newLink));
+        }
+
+        private static bool IsOverlapping(ChargeLink existingLink, ChargeLinkDto newLink)
+        {
+            // See https://stackoverflow.com/questions/325933/determine-whether-two-date-ranges-overlap
+            var isOverlapping = (existingLink.StartDateTime < newLink.EndDateTime || newLink.EndDateTime == null)
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/Dtos/ChargeLinksCommands/Validation/BusinessValidation/ValidationRules/ChargeLinksUpdateNotYetSupportedRule.cs:51]: nit, but in terms of evaluation I would feel safer with this (I guess that - at least in theory - Instant < null could return true):
```suggestion
            var isOverlapping = (newLink.EndDateTime == null || existingLink.StartDateTime < newLink.EndDateTime)
```

**Review by BjarkeMeier — APPROVED** [review]

**Pull request merged**